c10/core/InferenceMode.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+namespace c10 {
+
+// A RAII, thread local (!) guard that enables or disables inference mode upon
+// construction, and sets it back to the original value upon destruction.
+struct TORCH_API InferenceMode {
```

Comment:
If you turn on inference mode, and then you ship some work to a thread in a thread pool, inference mode won't apply inside that thread (because it's TLS). Because you added a new TLS for inference mode, even if dispatch key TLS is propagated, that inference mode TLS won't get propagated unless you add the logic in the relevant spots.

Reply:
We have a central TLS state system in c10 that is used to propagate states across threads.
Namely: https://github.com/pytorch/pytorch/blob/db5e5b439c454d657cfa8f08a096cf68e203f2a8/aten/src/ATen/ThreadLocalState.h#L11

You want to add a field there for inference mode and properly update the get and set functions.